### PR TITLE
[resotocore][feat] Automated Graph Snapshots

### PR DESCRIPTION
**Pull request by meln1k**
# Description

Do graph snapshots based on cron expression schedule.

# To-Dos

- [x] Add test coverage for new or updated functionality
- [x] Lint and test with `tox`
- [ ] Document new or updated functionality (someengineering/resoto.com#XXXX)

# Issues Fixed

- Fixes #XXXX

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://resoto.com/code-of-conduct).

### CODE REVIEW

**Comment by lloesche** [comment]
Let's rename the jobs to e.g.
```
resoto:snapshots:daily
```

And prohibit users to edit any jobs with `:` in their ID.

File: resotocore/resotocore/config/core_config_handler.py [hunk 1]
```diff
@@ -76,6 +81,22 @@ def validate_commands_config() -> Optional[Json]:
             else:
                 return {"error": "Expected a json object"}
 
+        def validate_snapshot_schedule() -> Optional[Json]:
+            config = value_in_path(task_data, ["config", ResotoCoreSnapshotsRoot])
+            if isinstance(config, dict):
+                config = {"snapshots": config}  # wrap in snapshots objects to match the dataclass
+                # try to read editable config, throws if there are errors
```
Comment (aquamatthias) [inline resotocore/resotocore/config/core_config_handler.py:88]: Wrong comment.

**Review by aquamatthias — COMMENTED** [review]
Thanks. A couple of requests:

- I see a lot of errors in my console:
```
Periodic function outdated_updates_cleaner caught an exception: [HTTP 400][ERR 1501] AQL: syntax error, unexpected integer number near '5-23T09_20_02Z_in_progress
```
- I was able to delete a job with `:`
```
> jobs delete resoto:snapshots:daily
Job resoto:snapshots:daily deleted.
```
- I could not search in any of the snapshots:
```
> graph=snapshot-resoto-foo-2023-05-23T09_20_02Z search is(graph_root)
```

File: resotocore/resotocore/graph_manager/graph_manager.py [hunk 2]
```diff
@@ -85,20 +86,20 @@ async def _on_config_updated(self, config_id: str, data: Json) -> None:
             await self.__setup_cleanup_old_snapshots_worker(snapshots_config)
 
             # cancel all existing snapshot jobs
-            existing_jobs = [job for job in await self.task_handler.list_jobs() if job.id.startswith("snapshot-")]
+            existing_jobs = [job for job in await self.task_handler.list_jobs() if job.id.startswith(job_prefix)]
             for job in existing_jobs:
                 await self.task_handler.delete_job(job.id)
 
             # schedule new snapshot jobs for the current graph
             for label, schedule in snapshots_config.snapshots.items():
```
Comment (lloesche) [inline resotocore/resotocore/graph_manager/graph_manager.py:94]: Is the label being cleaned? Consider the label being user-input and only allow [a-zA-Z0-9] as value.

**Comment by lloesche** [comment]
> ```
> > graph=snapshot-resoto-foo-2023-05-23T09_20_02Z search is(graph_root)
> ```

Could that be because of the missing backticks around collection names? I think we need to go through the entire resotocore codebase and wherever we use a collection or graph name in an AQL query surround it with backticks.

File: resotocore/resotocore/graph_manager/graph_manager.py (hunk 2 above)
Comment (meln1k) [reply resotocore/resotocore/graph_manager/graph_manager.py:94]: That's already the case.

**Review by aquamatthias — APPROVED** [review]
Thanks!

**Pull request merged**